Shopify/vscode-ruby-lsp
# Run the Ruby LSP from a generated custom Gemfile

## Description

Closes Shopify/ruby-lsp#1475
Closes Shopify/ruby-lsp#1536 (no longer needed)
Closes Shopify/ruby-lsp#1495 (no longer applicable)
Step towards Shopify/ruby-lsp#1496

Remove the need of adding the `ruby-lsp` to a project's main bundle by generating a custom `Gemfile`. The technique is basically to
- Create a `.ruby-lsp` folder
- Automatically gitignore it
- Add a `Gemfile` that evaluates the main `Gemfile` and adds the `ruby-lsp`
- Copy the `Gemfile.lock` to get the locked versions of RuboCop
- Execute the language server pointing `BUNDLE_GEMFILE` to our generated custom one

### Advantages

- Users don't have to add the Ruby LSP to the bundle
- RuboCop versions and extensions are still respected
- Users automatically get updates. Every time the LSP is started, we copy the `Gemfile.lock` and run `bundle install`
- Users can use the Ruby LSP with projects they don't control the dependencies (e.g.: a contributor to Rails or other open source projects)

## Manual testing

These changes are impactful, so please take the time to help manual test and verify that everything is working.

All scenarios involve
1. Switching to this branch
2. Yarn install
3. Start debug mode (F5)

#### On the ruby-lsp repo
1. Verify the LSP is working (test diagnostics or formatting)
2. Verify no `.ruby-lsp` folder is created. This is important because if it gets created it we'll have two conflicting versions of the gem (one coming from the custom gemfile and another one coming from the repo itself)

#### On a repo with ruby-lsp in the bundle

I suggest using Tapioca.

1. Verify you get prompted to remove the ruby-lsp from the bundle
2. Verify an ignored `.ruby-lsp` folder is added with Gemfile and Gemfile.lock
3. Verify the LSP is working despite not being a part of the bundle

#### On a repo that is not using the ruby-lsp yet

1. Verify you don't get prompted to remove the ruby-lsp from the bundle
2. Verify an ignored `.ruby-lsp` folder is added with Gemfile and Gemfile.lock
3. Verify the LSP is working despite not being a part of the bundle

#### On a repo that doesn't use rubocop

1. Verify an ignored `.ruby-lsp` folder is added with Gemfile and Gemfile.lock
2. Verify the LSP is working despite not being a part of the bundle
3. Verify diagnostics don't appear
4. Verify formatting works (with Syntax Tree)
